File: app/src/main/java/com/dicoding/moviecataloguerv/viewmodel/TvShowsViewModel.java
package com.dicoding.moviecataloguerv.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.dicoding.moviecataloguerv.data.Repository;
import com.dicoding.moviecataloguerv.data.source.remote.response.TvShowResponse;

public class TvShowsViewModel extends ViewModel {

    private LiveData<TvShowResponse> popularTv;
    private LiveData<TvShowResponse> nowPlayingTv;
    private LiveData<TvShowResponse> topRatedTv;
    private LiveData<TvShowResponse> newReleaseMovies;

    private Repository repository;

    public TvShowsViewModel(Repository repository) {
        this.repository = repository;
    }

    public void setPopularTv() {
        popularTv = new MutableLiveData<>();
        popularTv = repository.getTvShows("popular");
        Log.d("PopularTv", "Data Created");
    }

    public LiveData<TvShowResponse> getPopularTv() {
        if (popularTv == null) {
            setPopularTv();
        }
        return popularTv;
    }

    public void setOnAirTv() {
        nowPlayingTv = new MutableLiveData<>();
        nowPlayingTv = repository.getTvShows("on_the_air");
        Log.d("NowPlayingTv", "Data Created");
    }

    public LiveData<TvShowResponse> getOnAirTv() {
        if (nowPlayingTv == null) {
            setOnAirTv();
        }
        return nowPlayingTv;
    }

    public void setTopRatedTv() {
        topRatedTv = new MutableLiveData<>();
        topRatedTv = repository.getTvShows("top_rated");
        Log.d("TopRatedTv", "Data Created");
    }

    public LiveData<TvShowResponse> getTopRatedTv() {
        if (topRatedTv == null) {
            setTopRatedTv();
        }
        return topRatedTv;
    }

    public void setNewReleaseTv() {
        newReleaseMovies = new MutableLiveData<>();
        newReleaseMovies = repository.getNewReleaseTvShows();
        Log.d("NewRelease", "Fetched");
    }

    public LiveData<TvShowResponse> getNewReleaseTv() {
        if (newReleaseMovies == null) {
            setNewReleaseTv();
        }
        return newReleaseMovies;
    }
}

File: app/src/main/kotlin/com/ryanrvldo/moviecatalogue/ui/viewmodel/TvShowsViewModel.kt
package com.ryanrvldo.moviecatalogue.ui.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import com.ryanrvldo.moviecatalogue.data.Repository
import com.ryanrvldo.moviecatalogue.data.remote.response.TvShowResponse
import com.ryanrvldo.moviecatalogue.data.vo.Resource
import com.ryanrvldo.moviecatalogue.utils.Constants
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject

@HiltViewModel
class TvShowsViewModel @Inject constructor(
    private val repository: Repository,
) : ViewModel() {

    val popularTv: LiveData<Resource<TvShowResponse>> by lazy {
        repository.getTvShows(Constants.POPULAR_CATEGORY)
    }

    val nowPlayingTv: LiveData<Resource<TvShowResponse>> by lazy {
        repository.getTvShows(Constants.ON_AIR_CATEGORY)
    }

    val topRatedTv: LiveData<Resource<TvShowResponse>> by lazy {
        repository.getTvShows(Constants.TOP_RATED_CATEGORY)
    }
    val newReleaseTvShows: LiveData<TvShowResponse> by lazy {
        repository.getNewReleaseTvShows()
    }

    companion object {
        private const val TAG = "TvShowsVM"
    }

}
